### PR TITLE
Add OpenAPI specification

### PR DESCRIPTION
**Pull request by awesome-manuel**
OpenAPI is more precise than plain Markdown documentation of the APIs.

Signed-off-by: Manuel Stahl <manuel.stahl@awesome-technologies.de>

### CODE REVIEW

**Comment by nickvergessen** [comment]
What tool did you use to generate this?

File: docs/ocs-endpoint-v1.yaml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        label:
+          description: Translated short label of the action/button that should be presented to the user
+          type: string
+        link:
+          description: A link that should be followed when the action is performed/clicked
+          type: string
+        type:
+          description: HTTP method that should be used for the request against the link
+          type: string
+          enum: [GET, POST, DELETE]
```
Comment (nickvergessen) [inline docs/ocs-endpoint-v1.yaml]: missing PUT:
https://github.com/nextcloud/server/blob/ba87db3fccb40aa58d84d12932424c83a4db411f/lib/private/Notification/Action.php#L137

File: docs/ocs-endpoint-v2.yaml [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        label:
+          description: Translated short label of the action/button that should be presented to the user
+          type: string
+        link:
+          description: A link that should be followed when the action is performed/clicked
+          type: string
+        type:
+          description: HTTP method that should be used for the request against the link
+          type: string
+          enum: [GET, POST, DELETE]
```
Comment (nickvergessen) [inline docs/ocs-endpoint-v2.yaml]: missing put https://github.com/nextcloud/server/blob/ba87db3fccb40aa58d84d12932424c83a4db411f/lib/private/Notification/Action.php#L137

File: docs/ocs-endpoint-v2.yaml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      properties:
+        type:
+          description: A unique identifier for the object type
+          type: string
+        id:
+          description: A short identifier of the object on the server (int or string)
+          type: string
+        name:
+          description: A name which should be used in the visual representation
+          type: string
```
Comment (nickvergessen) [inline docs/ocs-endpoint-v2.yaml]: https://github.com/nextcloud/server/blob/master/lib/public/RichObjectStrings/Definitions.php contains all definitions. should we document the optional/type-depending properties too?

File: docs/ocs-endpoint-v2.yaml [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    Meta:
```
Comment (nickvergessen) [inline docs/ocs-endpoint-v2.yaml]: The order feels a bit weird, should we just "fix it" to the logical order?
Notifications > Meta > Notifcation > Rich... > Action

**Comment by awesome-manuel** [comment]
I used the [KaiZen Eclipse plugin](https://marketplace.eclipse.org/content/kaizen-openapi-editor), but the [Swagger online editor](https://editor.swagger.io/) works as well.

File: docs/ocs-endpoint-v2.yaml (hunk 4 above)
Comment (awesome-manuel) [reply docs/ocs-endpoint-v2.yaml]: Could be done, or maybe alphabetical or more convenient.

File: docs/ocs-endpoint-v2.yaml (hunk 3 above)
Comment (awesome-manuel) [reply docs/ocs-endpoint-v2.yaml]: Not sure if I understand this correctly: "id", "name", "server" are possible values for RichStringParameter::id ?

File: docs/ocs-endpoint-v2.yaml [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                $ref: "#/components/schemas/Notification"
+    delete:
+      summary: Delete a notification for a user
+      tags:
+        - notifications
+      parameters:
+        - name: id
+          in: path
+          required: true
+          description: The id of the notification to retrieve
```
[the remark this answers is not in the document]
Comment (nickvergessen) [reply docs/ocs-endpoint-v2.yaml]: github now showing the comment on the wrong lines, of course its retrieve here

File: docs/ocs-endpoint-v2.yaml (hunk 3 above)
Comment (nickvergessen) [reply docs/ocs-endpoint-v2.yaml]: >  "addressbook", "addressbook-contact", "announcement" are possible values for RichStringParameter::type

all keys, are possible types, yes.

but depending on the type, they dont have only "type, id and name" (those 3 are required for all), but also additionally, e.g. link for announcement https://github.com/nextcloud/server/blob/master/lib/public/RichObjectStrings/Definitions.php#L95
link for circles https://github.com/nextcloud/server/blob/master/lib/public/RichObjectStrings/Definitions.php#L201
call-type for call https://github.com/nextcloud/server/blob/master/lib/public/RichObjectStrings/Definitions.php#L176
and more

Comment (awesome-manuel) [reply docs/ocs-endpoint-v2.yaml]: Alright, the dependencies are not so easy to model. Can be done, but needs a lot of redundant lines. For now I'd just add all optional values.

**Comment by nickvergessen** [comment]
I'm closing this for now since it's quite outdated and old.

I think we need something that is generated better from the actual code, otherwise it is doomed to be outdated soon.

**Pull request closed without merging**